File: mcp_server/scripts/smoke.js
import { spawn } from "node:child_process";
import { once } from "node:events";

function frame(message) {
  const body = Buffer.from(JSON.stringify(message), "utf8");
  return `Content-Length: ${body.length}\r\n\r\n${body}`;
}

const child = spawn(process.execPath, ["src/index.js"], { cwd: new URL("..", import.meta.url), stdio: ["pipe", "pipe", "inherit"] });
let output = "";
child.stdout.on("data", (chunk) => {
  output += chunk.toString("utf8");
});

child.stdin.write(frame({ jsonrpc: "2.0", id: 1, method: "initialize", params: {} }));
child.stdin.write(frame({ jsonrpc: "2.0", id: 2, method: "tools/list", params: {} }));
child.stdin.end();

await once(child, "exit");

if (!output.includes("fitlog-context-mcp") || !output.includes("get_daily_meals") || !output.includes("create_strategy")) {
  console.error(output);
  process.exit(1);
}

console.log("MCP smoke test passed");

File: backend/app/mcp/smoke.py
from __future__ import annotations

import json
import subprocess
import sys


def frame(message: dict[str, object]) -> bytes:
    body = json.dumps(message).encode("utf-8")
    return f"Content-Length: {len(body)}\r\n\r\n".encode("ascii") + body


def main() -> int:
    process = subprocess.run(
        [sys.executable, "-m", "app.mcp.server"],
        input=b"".join(
            [
                frame({"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": {}}),
                frame({"jsonrpc": "2.0", "id": 2, "method": "tools/list", "params": {}}),
            ]
        ),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=False,
    )
    output = process.stdout.decode("utf-8", errors="replace")
    if process.returncode != 0:
        sys.stderr.write(process.stderr.decode("utf-8", errors="replace"))
        return process.returncode
    if "fitlog-context-mcp" not in output or "get_daily_meals" not in output or "create_strategy" not in output:
        sys.stderr.write(output)
        return 1
    print("Python MCP smoke test passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

